src/random-walk.js

```diff
@@ -59,70 +58,67 @@ class RandomWalk {
    * @returns {void}
    */
   stop () {
-    clearTimeout(this._timeoutId)
-    this._timeoutId = null
+    this._running = false
     this._controller && this._controller.abort()
   }
 
   /**
-   * Run function `walk` on every `interval` ms
-   * @param {function(callback)} walk The function to execute on `interval`
-   * @param {number} interval The interval to run on in ms
+   * Run function `randomWalk._walk` on every `options.interval` ms
    *
    * @private
    */
-  _runPeriodically (walk, interval) {
-    this._timeoutId = setTimeout(() => {
-      walk((nextInterval) => {
-        // Schedule next
-        this._runPeriodically(walk, nextInterval)
-      })
-    }, interval)
+  async _runPeriodically () {
+    while (true) {
+      if (!this._running) return
+      try {
+        await this._walk(this._options.queriesPerPeriod, this._options.timeout)
+      } catch (err) {
+        this._kadDHT._log.error('random-walk:error', err)
+      }
+      // Each subsequent walk should run on a `this._options.interval` interval
+      await new Promise(resolve => setTimeout(resolve, this._options.interval))
```

Comment:
Can we use `Promise.race()` here to detect when the random walk has been stopped? Otherwise we may need to wait a full interval when shutting down the DHT. (Note that you may need to use an event that gets fired on stop)

Reply:
@dirkmc are you just concerned about the timeout keeping a node process alive?
if so we can [unref it](https://nodejs.org/api/timers.html#timers_timeout_unref)
otherwise it shouldnt be an issue

I guess one problem could be if `stop` and `start` are called during the interval, we'll end up with two loops running...

Reply:
That's true I guess you can just keep a reference and call clearTimeout()